ci/scripts/utils/rocotostat.py

```diff
@@ -39,60 +42,138 @@ def input_args():
+    command, runs it, and processes its output to return a dictionary with the count
+    of each status case.
+
+    Input:
+    rocotostat - The rocotostat command.
+
+    Output:
+    rocoto_status - A dictionary with the count of each status case.
+    """
+
+    rocotostat = copy.deepcopy(rocotostat)
+    rocotostat.add_default_arg('--all')
+
+    rocotostat_output = rocotostat(output=str)
+    rocotostat_output = rocotostat_output.splitlines()[1:]
+    rocotostat_output = [line.split()[0:4] for line in rocotostat_output]
+    rocotostat_output = [line for line in rocotostat_output if len(line) != 1]
 
     status_cases = ['SUCCEEDED', 'FAIL', 'DEAD', 'RUNNING', 'SUBMITTING', 'QUEUED']
+
+    rocoto_status = {}
+    status_counts = Counter(case for sublist in rocotostat_output for case in sublist)
     for case in status_cases:
-        rocoto_status[case] = sum([sublist.count(case) for sublist in rocotostat_output_all])
+        rocoto_status[case] = status_counts[case]
 
     return rocoto_status
 
 
+def is_done(rocoto_status):
+    """
+    is_done Check if all cycles are done.
+
+    is_done(rocoto_status) checks if the total number of cycles equals the number of
+    done cycles in the rocoto_status dictionary.
+
+    Input:
+    rocoto_status - A dictionary with the count of each status case.
+
+    Output:
+    boolean - True if all cycles are done, False otherwise.
+    """
+
+    if rocoto_status['CYCLES_TOTAL'] == rocoto_status['CYCLES_DONE']:
+        return True
+    else:
+        return False
+
+
+def is_stalled(rocoto_status):
+    """
+    is_stalled Check if all cycles are stalled.
+
+    is_stalled(rocoto_status) checks if all cycles are stalled by verifying if
+    there are no jobs that are RUNNING, SUBMITTING, or QUEUED.
+
+    Input:
+    rocoto_status - A dictionary with the count of each status case.
+
+    Output:
+    boolean - True if all cycles are stalled, False otherwise.
+    """
+
+    if rocoto_status['RUNNING'] + rocoto_status['SUBMITTING'] + rocoto_status['QUEUED'] == 0:
+        return True
+    else:
+        return False
+
+
 if __name__ == '__main__':
+    """
+    main Execute the script.
+
+    main() parses the input arguments, checks if the rocotostat command is available,
+    adds default arguments to the rocotostat command, and runs it and reports
+    out to stdout spcific information of rocoto workflow.
+    """
 
     args = input_args()
 
+    try:
+        rocotostat = which("rocotostat")
+    except CommandNotFoundError:
+        logger.exception("rocotostat not found in PATH")
+        raise CommandNotFoundError("rocotostat not found in PATH")
+
+    rocotostat.add_default_arg(['-w', os.path.abspath(args.w.name), '-d', os.path.abspath(args.d.name)])
+
     error_return = 0
-    rocoto_status = rocoto_statcount()
+    rocoto_status = rocoto_statcount(rocotostat)
+    rocoto_status.update(rocotostat_summary(rocotostat))
```

Comment:
```suggestion
    try:
        rocoto_status = rocoto_statcount(rocotostat)
        rocoto_status.update(rocotostat_summary(rocotostat))
    except ProcessError:
        # rocotostat sometimes fails. This does not mean the case is necessarily faulty.
        # Try again later and hope it is resolved. Repeated failures will have to be killed manually.
        logger.exception("Rocotostat failed. This may be a temporary error. Will try again later.")
        sys.exit(error_return)        
```

Reply:
I will update the PR so that we have it fail and message appropriately in the framework without the logging.

Reply:
attempt_multiple_times works great!!!